### ISSUE TITLE
Linux, OpenGL + GLFW - Random memory errors on application exit.

### ISSUE THREAD

**Issue by 1Kuso4ek1** [LukasBanana/LLGL#138]
So recently I've created a small app with LLGL using LLGL::Window as a main surface, but my goal was to use GLFW and ImGui.</br>
This is my project: https://github.com/1Kuso4ek1/LLGLTest</br>
![image](https://github.com/user-attachments/assets/eb782de0-4afe-43c9-be7f-74eed955ffea)</br>
The weird part starts when I exit the app. Everything can be normal even for multiple runs, but then I randomly get `corrupted size vs. prev_size`, `double free or corruption (!prev)` or `corrupted double-linked list`.</br>
I've tried to debug with no actual success (found out that errors are related to the context, e. g. deleting it), but through trial and error I figured out that removing all **pipeline** related code (CreatePipelineState, RenderPass in [CubeApp.cpp](https://github.com/1Kuso4ek1/LLGLTest/blob/master/src/CubeApp.cpp)) fixes the issue, and I really don't understand why (but I guess that the issue is in my code), **so any help will be appreciated**</br>
OS: Debian sid with Gnome 47 on Wayland (GLFW runs under X11)
GPU: Integrated AMD Radeon Vega 7

**Comment by LukasBanana** [comment]
Hi, I see you're using GLFW not only to create a window but also to use its GL context. LLGL needs to manage its own GL context, so you should always use `swapChain->Present()` instead of `glfwSwapBuffers()`. Same goes for `glfwMakeContextCurrent()`, this will likely confuse LLGL as it doesn't know you made an external GL context the current one. It seems GLFW will always create a GL context whether you use it or not, but you should use GLFW solely for its windowing functionality.

If you need to access the native GL context, you can use [`LLGL::RenderSystem::GetNativeHandle()`](https://github.com/LukasBanana/LLGL/blob/master/include/LLGL/RenderSystem.h#L658) (despite the docs saying it's not supported for OpenGL, you can ignore that part).

This might not be the final fix, but let me know if that improves your situation.

I also see you're planning to do some multithreading tests but you don't allocate any threads yet. Just keep in mind that OpenGL was never designed for multi-threading, so all your commands must be scheduled on the main thread, or to be more precise, on the same thread that owns the GL context. So whenever you're planning to submit draw commands in parallel, make sure to record them with a deferred command buffer (as opposed to an immediate-submit buffer) and submit them on the main thread.

**Comment by 1Kuso4ek1** [comment]
Hello! Thank you for the answer!</br>

1. Using only `swapChain->Present()` I can't **disable** VSync for some reason, so I'll stick with `glfwSwapBuffers` for now.
2. As for multithreading, thank you for telling about deferred CB, it will come in handy. For now I only use threads to asynchronously load textures - `stbi_load` on a new thread, uploading to the gpu on the main thread, so it works fine I guess.
3. Seems like the issue was in the destructors call order or smth like that xd. Here's how I solved it:
```c++
// Renderer.cpp
void Renderer::Unload()
{
    LLGL::RenderSystem::Unload(std::move(renderSystem));
}
```
```c++
// CubeApp.cpp
CubeApp::~CubeApp()
{
    DestroyImGui();

    dev::Renderer::Get().Unload();
}
```
That's all, huh... I'm sorry for wasting your time :3</br>
Thanks again.

**Comment by LukasBanana** [comment]
> I can't disable VSync for some reason, so I'll stick with glfwSwapBuffers for now.

Do you mean [`LLGL::SwapChain::SetVsyncInterval()`](https://github.com/LukasBanana/LLGL/blob/master/include/LLGL/SwapChain.h#L170) does not work? Using `glfwSwapBuffers()` will be undefined behavior nonetheless. I.e. other issues might occur later down the line.

> That's all, huh... I'm sorry for wasting your time :3

No worries and thanks for reaching out.

**Comment by 1Kuso4ek1** [comment]
> Do you mean [`LLGL::SwapChain::SetVsyncInterval()`](https://github.com/LukasBanana/LLGL/blob/master/include/LLGL/SwapChain.h#L170) does not work?

Yes, indeed. Setting it to zero does nothing (60 fps lock), while everything >= 1 works as expected. This is also on of the reasons for me to switch from LLGL::Window to GLFW, besides strange, buggy fullscreen mode.

**Comment by LukasBanana** [comment]
Can you please check if [`glXSwapIntervalSGI`](https://github.com/LukasBanana/LLGL/blob/master/sources/Renderer/OpenGL/Platform/Linux/LinuxGLContext.cpp#L184) ever gets called there or otherwise share what the following command prints on your machine?
```
glxinfo | grep GL_
```
I might have to add two more GL extensions to support other versions of the GLX swap interval.

**Comment by 1Kuso4ek1** [comment]
It's being called, but this whole `SetSwapInterval` function returns `false`.</br>
![image](https://github.com/user-attachments/assets/5aa1156b-8c4f-4f25-973d-ecf1dc36e37d)</br>
`glxinfo | grep swap_control` output:
```
    GLX_EXT_import_context, GLX_EXT_no_config_context, GLX_EXT_swap_control, 
    GLX_EXT_swap_control_tear, GLX_EXT_texture_from_pixmap, 
    GLX_MESA_swap_control, GLX_NV_float_buffer, GLX_OML_sync_control, 
    GLX_SGI_swap_control, GLX_SGI_video_sync
    GLX_EXT_framebuffer_sRGB, GLX_EXT_no_config_context, GLX_EXT_swap_control, 
    GLX_EXT_swap_control_tear, GLX_EXT_texture_from_pixmap, 
    GLX_MESA_gl_interop, GLX_MESA_query_renderer, GLX_MESA_swap_control,
```

**Comment by LukasBanana** [comment]
Thanks for reporting this. I was able to confirm this issue in my Linux VM. I added support for the other two swap-control extensions `GLX_MESA_swap_control` and `GLX_EXT_swap_control` which both work, I can see a much higher frame rate with `interval=0`.

**Comment by 1Kuso4ek1** [comment]
Oh, thank you, it works perfect now! I get like 5 000 fps instead of 300 with swap buffers :+1:
